Rewrite and extend computer graphics notes to follow a more conventional style (#3)
VIK-CE-Notes/vik-ce-notes.github.io: docs/notes/sem4/computer_graphics/lectures/5.md

# 2D képszintézis

Feltételezzük, hogy a virtuális világunk a kétdimenziós Euklideszi sík. Hogyan tudjuk ezt a világot megjeleníteni a képernyőn?

> Vannak objektumjaink, amiknek tudjuk a helyét a virtuális világban. Szükségünk lesz egy kamerára, ami azt jelöli majd ki, hogy a virtuális világból mennyit látunk (hol vagyunk, merre nézünk, stb). A végső fénykép pixelekből áll, szóval minden pixelről el kell dönteni, hogy milyen színű legyen. Ha minden objektumhoz rendelünk egy színt, aztán pedig ez az objektum feltűnik egy adott pixelen, akkor tudjuk, hogy annak a pixelnek milyen színűnek kell lennie. (Ha több objektum is verseng egy adott pixelért, azaz van overlap, akkor explicit prioritást használunk, azaz konkrétan van egy szám minden objektumhoz, ami megmondja, hogy milyen sorrendben rajzoljuk ki őket)

Reviewer: ```suggestion
Vannak objektumjaink, amiknek tudjuk a helyét a virtuális világban. Szükségünk lesz egy kamerára, ami azt jelöli majd ki, hogy a virtuális világból mennyit látunk (hol vagyunk, merre nézünk stb.). A végső fénykép pixelekből áll, tehát minden pixelről el kell dönteni, hogy milyen színű legyen. Ha minden objektumhoz rendelünk egy színt, és ez az objektum feltűnik egy adott pixelen, akkor tudjuk, hogy annak a pixelnek milyen színűnek kell lennie. (Ha több objektum is verseng egy adott pixelért, azaz van átfedés, akkor explicit prioritást használunk, azaz tartozik egy szám minden objektumhoz, ami megmondja, hogy milyen sorrendben rajzoljuk ki őket.)
```

Hogyan tudunk kapcsolatot teremteni az objektumaink és a képernyő között? Két irányból közelíthetjük meg:

- pixelvezértel (pixel $\rightarrow$ objektumok)
- objektumvezérelt (objektumok $\rightarrow$ pixel)

Az objektumvezérelt gyakran hatékonyabb, mint a pixelvezérelt, azt fogjuk preferálni.

## Pixelvezérelt képszintézis

> Induljunk a pixelek felől. A képernyőn véges számú pixel van, tehát elég annyit csinálnunk, hogy minden egyes pixelben eldöntjük, hogy melyik objektum látszódik benne. Transzformáljuk a pixeleket világkoordináta-rendszerbe, és ezután pedig amelyik objektumra landolunk, az van abban a pixelben (figyelembe véve a prioritásukat).

Reviewer: ```suggestion
Induljunk a pixelek felől. A képernyőn véges számú pixel van, tehát elég annyit csinálnunk, hogy minden egyes pixelben eldöntjük, hogy melyik objektum látszódik benne. Transzformáljuk a pixeleket világkoordináta-rendszerbe, majd amelyik objektumra esik az adott pixel, az fogja meghatározni a színét (figyelembe véve a prioritásukat).
```

Ha diagonálok mentén vágjuk fel az alakzatunkat, akkor minden vágás után garantáltan két kisebb csúcsszámú alakzatot kapunk, tehát véges lépés után csak háromszögeink maradnak majd. Gondoljunk bele, hogy ez közel sem garantált, ha random vagdossuk az alakzatunkat, sőt, néha még növelhetjük is a csúcsszámot.

Konvex sokszögek esetén minden nem szomszédos csúcspár egy diagonált alkot, tehát itt hatékony algoritmust ($O(n)$) adhatunk az alakzat háromszögekre bontására. Konkáv sokszögekre ez már nem igaz, tehát itt lényegesen bonyolultabb algoritmust kell majd használnunk ($O(n^3)$).

Felmerülhet a kérdés, hogy mi a garancia arra, hogy minden konkáv sokszögben van annyi diagonál, hogy felbontható legyen háromszögekre. Az alábbi tétel azt mondja ki, hogy minden _egyszerű_ sokszögre ez igaz lesz.

!!! info 5.1 Tétel (Egyszerű sokszögek diagonáljának létezése)
    Minden legalább $4$ csúcsú <u>__egyszerű__</u> sokszögnek van diagonálja, azaz mindegyik felbontható diagonálok mentén.

Egy sokszög pedig akkor _egyszerű_, ha az oldalai nem keresztezik egymást és "megrajzolható a ceruza felemelése nélkül".

??? abstract Bizonyítás

>     Keressük meg a sokszögünk egy szélső csúcsát, mondjuk az $x$ irányba (tehát a legnagyobb $x$ koordinátájú csúcs). Vegyük a két szomszédját és kössük őket össze. Ekkor két dolog történhet:

Reviewer: ```suggestion
    Keressük meg a sokszögünk egy szélső csúcsát, mondjuk az $x$ irányban (tehát a legnagyobb $x$ koordinátájú csúcs). Vegyük a két szomszédját és kössük őket össze. Ekkor két dolog történhet:
```

not sure tho

#### Implementáció (vektorizálás)

Nézzünk egy példát arra, hogy ha vannak 2D-s objektumaink, akkor azokat hogyan tudjuk vektorizálni. Ha van egy Bézier görbe osztályunk, akkor fel kell bontanunk kicsi szakaszokra. Ehhez elég, ha mintavételezzük a görbe pontjait elég sűrűn:

```cpp

const int nTessVertices = 100;

//...

vector<vec2> GenVertexData() { // vectorization
    vector<vec2> vertices;
    for(int i = 0; i <= nTessVertices; i++) {
        float t = (float)i / nTessVertices;
        vertices.push_back(r(t));
    }
    return vertices;
}
```

Ez a függvény $100$ szakaszra bontja a görbénket, ahol az `r(t)` függvény a görbe egyik pontját adja `t` futóváltozó szerint.

Később majd látunk egy komplettebb példát, a modell és kamera transzformációk után.

## Modell és kamera transzformáció

A vektorizációt követően már csak pontjaink, szakaszaink, és háromszögeink vannak. Ez nekünk azért jó, mert tudunk olyan transzformációkat találni, amik ezeket az alakzatokat "megtartják", azaz háromszöget háromszögbe, szakaszt pedig szakaszba transzformálnak (homogén lineáris transzformációk).

> Magukat a transzformációkat elég csak a csúcsokra végrehajtani, nem kell például egy szakasz minden (végtelen sok) pontjára ezt megtenni. Ez nem magától értetődő, direkt azért használunk homogén lineáris transzformációkat, hogy ez a tulajdonság fenn álljon, hiszen míg matematikailag lehet végtelen sok ponttal dolgozni, a GPU-nak ez nehezen menne.

Reviewer: ```suggestion
Magukat a transzformációkat elég csak a csúcsokra végrehajtani, nem kell például egy szakasz minden (végtelen sok) pontjára ezt megtenni. Ez nem magától értetődő, direkt azért használunk homogén lineáris transzformációkat, hogy ez a tulajdonság fennálljon, hiszen míg matematikailag lehet végtelen sok ponttal dolgozni, a GPU-nak ez nehezen menne.
```

    vec2 wCenter; // center in world coords
    vec2 wSize; // width and height in world coords
   public:
    // view matrix
    mat4 V() { return translate(vec3(-wCenter.x, -wCenter.y, 0)); }

    // projection matrix
    mat4 P() {
        return scale(vec3(2/wSize.x, 2/wSize.y, 1));
    }

    // inverse view matrix
    mat4 Vinv() { 
        return translate(vec3(wCenter.x, wCenter.y, 0));
    }

    // inverse projection matrix
    mat4 Pinv() { 
        return scale(vec3(wSize.x/2, wSize.y/2, 1));
    }

    void Zoom(float s) { wSize = wSize * s; }
    void Pan(vec2 t) { wCenter = wCenter + t; }
};
```

Itt a `wCenter` és `wSize` azok világkoordinátákban értelmezendők. Mivel sok koordináta-rendszeren átmegyünk a transzformációk során, ezért jó programozási gyakorlat, hogy a változók nevében valahogy jelezzük, hogy azokat éppen milyen koordináta-rendszerben kell értelmezni.

<!-- NOTE: ötletem sincs, hogy ez a mondat mit jelent, de az előadásvideóban benne volt. nem fejti ki ott sem. -->
Megjelenik a view és projection mátrixok inverze is, ezekre a bemeneti csővezeték ágán lesz szükségünk.

### Összegzés

<!-- NOTE: ide jó lenne több magyarázat az MVP mátrixról -->

> Mivel a modell, view és projection transzformációk mind homogén lineáris transzformációk, ezért összevonhatjuk őket mátrixszorzással. Ennek a mátrixnak `MVP` mátrix a neve, és így áll elő:

Reviewer: ```suggestion
Mivel a modell, view és projection transzformációk mind homogén lineáris transzformációk, ezért összevonhatjuk őket mátrixszorzással. Ez az ún. `MVP` mátrix, és az alábbi módon állítható elő:
```

        mat4 M = translate(pos) * rotate(phi, vec3(0, 0, 1)) * scale(scaling);
        mat4 MVP = camera.P() * camera.V() * M;
        gpuProgram->setUniform(MVP, "MVP");
        Draw(gpuProgram, type, color);
    }
};
```

A `Geometry` osztály pontos implementációja most annyira nem érdekes, a `framework.h` része. A számunkra releváns része az, hogy egy `Vtx()` függvénnyel tudjuk megkapni azt a tömböt, amit fel fogunk tölteni a GPU-ra (az `updateGPU()` függvénnyel).

Vizsgáljuk meg, hogy mit csinálnak az egyes függvények:

- `GenVertexData()`: ez a függvény felel a vektorizációért. Mivel ez minden objektumra eltérő, ezért ez egy tisztán virtuális függvény.
- `update()`: ez a függvény feltölti a GPU-ra az objektumot. Ehhez először vektorizál, utána pedig VAO-kba és VBO-kba rakja a pontjainkat.
- `Draw()`: kirajzoljuk az objektumot. Ehhez a már megismert `Camera2D` osztály hívjuk segítségül. A vektorizáció után a modell, view és projection transzformációkat kell elvégezni.
    - először előállítjuk az `M` mátrixot, mégpedig a korábban tanult alakban: eltolás $\ast$ forgatás $\ast$ skálázás
    - ezután a kameránkból kinyert `V` és `P` mátrixokkal előállítjuk az `MVP` mátrixot

>     - végezetül pedig feltöltjük a mátrixunkat a GPU-ra mind uniform változót, és meghívjuk a `Geometry` osztály kirajzoló függvényét, ami az OpenGL-es hívásokat ténylegesen megteszi.

Reviewer: ```suggestion
    - végezetül pedig feltöltjük a mátrixunkat a GPU-ra, mint uniform változót, és meghívjuk a `Geometry` osztály kirajzoló függvényét, ami már az OpenGL függvényeit hívja meg.
```

Két shadert kell még megírnunk:

- vertex shader, ami elvégzi az MVP transzformációt,
- fragment shader, ami beállítja a pixelek színét

#### Vertex shader

Mivel a `Draw` függvény feltöltötte az MVP mátrixot mint uniform változót, ezért a vertex shaderben arra hivatkozunk, és szimplán minden pontot aszerint transzformálunk (ami után kibővítettük négydimenziós vektorra):

```glsl
uniform mat4 MVP;
layout(location = 0) in vec2 vertexPosition;

void main() {
    gl_Position = MVP * vec4(vertexPosition.x, vertexPosition.y, 0, 1);
}
```

#### Fragment shader

A fragment shader kap egy uniform változón keresztül egy színt (ezt a `Geometry` osztály `Draw` függvénye állítja be, az `Object` osztály amikor meghívja, akkor paraméterként ad egy `vec3 color` változót). A shaderünk ezt a színt minden pixelre beállítja, tehát ha például a `BezierCurve` osztályunkban valahol beállítjuk, hogy piros színe legyen, akkor minden pixel, ami a görbét előállítja, az piros lesz:

```glsl
uniform vec3 color;
out vec4 fragmentColor;

void main() {
    fragmentColor = vec4(color.x, color.y, color.z, 1);
}
```

> Ha ezt mind leprogramoznánk, és írnánk hozzá egy minimális főprogramot (`glApp` leszármazott) ami létrehozza az objektumainkat, akkor tényleg megjelenne a görbe a képernyőn.

Reviewer: ```suggestion
Ha ezt mind leprogramoznánk, és írnánk hozzá egy minimális főprogramot (`glApp` leszármazott), ami létrehozza az objektumainkat, akkor valóban megjelenne a görbe a képernyőn.
```

$$
z_{\text{pix}} = (z_{\text{ndc}} + 1) / 2
$$

## Raszterizáció

Eddig végig primitívekkel, és azok csúcsaival dolgoztunk, viszont a raszterizáció során már pixelekkel fogunk. Ez első ránézésre nem tűnik nagy változásnak, viszont vizsgáljuk meg, hogy eddig hány műveletbe kerülne mondjuk egy háromszöget transzformálni, és vágni.

Egy háromszögnek három csúcsa van, és csúcsonként a transzformációk körülbelül hat szorzás és hat összeadásba, a vágás pedig négy összehasonlításba kerül. Ezt három csúcsra megcsinálni még annyira nem hangzik vészesnek, viszont menjünk tovább a raszterizációra. Egy háromszög hány pixelből áll? Hát ez így egy értelmetlen kérdés, függ a háromszög méretétől, függ a monitor felbontásától, függ az ablak méretétől, stb..., viszont az biztos, hogy messze meghaladja a csúcsokon végzett műveletek számának nagyságrendjét.

Ha valós idejű animációt szeretnénk, akkor (minimum) harmincszor újra kell rajzolnunk a képernyőt másodpercenként. Egy standard, full HD monitornak $1920 \times 1080 = 2\,073\,600$ pixele van (4K monitoroknak pedig már $8\,294\,400$).

> Ha másodpercenként harmincszor akarunk kétmillió pixelt kirajzolni, akkor pixelenként $15$ ns jut átlagosan. Érezzük, hogy itt már sokkal nagyobb teljesítményű algoritmusokat kell majd alkotnunk, olyanokat, amiket szinkron sorrendi hálózatokból meg tudunk alkotni.

Reviewer: ```suggestion
Ha másodpercenként harmincszor szeretnénk kétmillió pixelt kirajzolni, akkor pixelenként $15$ ns jut átlagosan. Érezzük, hogy itt már sokkal nagyobb teljesítményű algoritmusokat kell majd alkotnunk, olyanokat, amiket szinkron sorrendi hálózatokból meg tudunk alkotni.
```

"meg tudunk alkotni" kicsit ehh de nem jut eszembe jobb most

### Szakaszok

Egy szakasz végtelenül vékony, és összefüggő alakzat. Ahhoz, hogy ezt a két jellemzőt megtartsuk amikor pixelekre alakítjuk a szakaszt muszáj lesz pixeloszloponként csak egy (vagy nagyon kevés) pixelt kiszínezni, így vékony marad, de összefüggő is lesz.

Egy naiv megközelítés lehetne az, hogy hát az egyenes egyenlete $y = mx+b$, ez alapján növeljük $x$-et és az $y$ értéket mindig kerekítve megkapjuk a pixelt, amit ki kell színezni. Ez bár működik, de lassú, minden pont kiszínezéséhez egy szorzást meg egy összeadást kell elvégezni és kerekíteni. Ha már csak pár tízezer pixelből állna a szakaszunk, akkor is használhatatlan lenne.

A megoldás az inkrementális elv. Egyszer kiszámoljuk az egyenes $m$ meredekségét és mindig csak hozzáadjuk az előző ponthoz:

$$
\begin{align*}
y(x) &= mx + b \\
y(x+1) &= m(x+1) + b = m + mx + b = m + y(x) \\
\implies y(x) &= y(x-1) + m
\end{align*}
$$

Itt még mindig kerekíteni kell, de megspórulunk sok felesleges szorzást.

> A kerekítés gyorsítása érdekében sok hardware megoldás született, az egyik ilyen a fixpontos inkrementális számítási elv:

Reviewer: ```suggestion
A kerekítés gyorsítása érdekében sok hardver megoldás született, az egyik ilyen a fixpontos inkrementális számítási elv:
```

        y = y + m;
    }
}
```

Ez ugyanúgy inkrementális, csak a bit shift-el végzi el a kerekítést, amivel már nagyon gyors. Ha ezt egy [DDA](https://en.wikipedia.org/wiki/Digital_differential_analyzer_(graphics_algorithm)) szakaszkirajzoló hardver végzi el, amivel még sokkal gyorsabb (3 regiszter, meg 1 összegző).

### Háromszögek

A háromszögeknek már van területe is, szóval el kell tudni dönteni sok pontról, hogy tartalmazza-e őket a háromszög. A naiv megközelítés az lenne, ha fognánk a már korábban megismert tartalmazási tesztet (félegyenes elindítása, metszéspontok megállapítása) és ezt minden pixelre egymás függetlenül elvégeznénk. Bár működne, érezzük, hogy kell lennie ennél hatékonyabb módszernek.

A fő felismerésünk az, hogy ha egy tulajdonság igaz egy pixelre, akkor nagy eséllyel a szomszédjaira is, hacsak valamilyen határt át nem lépünk. Csoportosítsuk a pixeleinkát _pásztákba_, azaz hosszú vízszintes sorokba. Ekkor egy pásztán belül az elején sok pixel nem lesz benne a háromszögben, aztán lesz egy határpont, ahol átbillenünk, és hirtelen minden pixel a háromszögön belül van, aztán meg egy második határpont, ami után már egyik pixel sem lesz a háromszögön belül.

> Nézzük a konkrét algoritmust: messük el a háromszögünket vízszintes pásztákkal, és számoljuk ki a két metszéspont $x$ koordinátáit (az $y$ ugyan az lenne, hiszen vízszintes az egyenesünk). Ekkor csak eközött a két $x$ koordináta között kell kiszínezni a pixeleket, nem kell semmi extra műveletet elvégezni a pásztára, ha ez a két érték megvan már.

Reviewer: a messük egy létező szó, im flabbergasted
```suggestion
Nézzük a konkrét algoritmust: messük el a háromszögünket vízszintes pásztákkal, és számoljuk ki a két metszéspont $x$ koordinátáit (az $y$ ugyan az lenne, hiszen az egyenesünk vízszintes). Ekkor csak eközött a két $x$ koordináta között kell kiszínezni a pixeleket, nem kell semmi extra műveletet elvégezni a pásztára, ha ez a két érték megvan már.
```

        - Magyarázat: egy fagráf jön létre, ami körmentes
    - [ ] Minden több mint három csúcsú sokszögnek van legalább két füle.
        - Magyarázat: nem, ha nem egyszerű akkor nincs

---

!!! question 8\. Melyek az alábbiak közül affin transzformációk?
    - Az $(1, 3)$ pont körüli forgatás
    - Helyben hagyás
    - $x$ tengely mentén végrehajtott skálázás
    - Origóra tükrözés
    - Nyírás: $x'=x$; $y'= y+ax$
    - Eltolás
    - $x$ tengelyre vetítés

??? tip Megoldás
    Mindegyik az, az inverzió lenne a kivétel, de az nem szerepelt.

---

!!! question 9\. Az alábbi 2D homogén lineáris transzformációs mátrix hány fokos elforgatást valósít meg:
    $$
    \begin{bmatrix}
    0 & -1 & 0 \\
    -1 & 0 & 0 \\
    0 & 0 & 1
    \end{bmatrix}
    $$

    Nyilván végtelen sok helyes válasz van, amelyek $360$ fok egész számú többszörösében különböznek. A minimális abszolút értékű megoldást várjuk.

??? tip Megoldás
    $\cos(\varphi) = 0$ és $\sin(\varphi) = -1 \implies \varphi = -90$

---

!!! question 10\. Hány füle van az alábbi 5-szögnek?
    ![quiz](../img/chapter_5/5_kviz_csillag.png){width=100%}

??? tip Megoldás

>     Mivel nincs diagonálja, ezért füle sincs (mert a fül az amikor egy pont szomszédai összeköthetők diagonállal).

Reviewer: ```suggestion
    Mivel nincs diagonálja, ezért füle sincs (mert a fül az, amikor egy pont szomszédai összeköthetők diagonállal).
```
